Amazon.Ion.ObjectMapper/IonSerializer.cs

```diff
@@ -172,87 +175,101 @@ public void Serialize<T>(IIonWriter writer, T item)
+                serializer.Serialize(writer, Convert.ToDecimal(item));
                 return;
             }
 
-            if (item is BigDecimal)
+            if (type.IsAssignableTo(typeof(BigDecimal)))
             {
-                new IonBigDecimalSerializer().Serialize(writer, (BigDecimal)(object)item);
+                var serializer = this.GetCustomPrimitiveSerializer<BigDecimal>() ?? new IonBigDecimalSerializer();
+                serializer.Serialize(writer, (BigDecimal)(object)item);
                 return;
             }
 
-            if (item is byte[])
+            if (type.IsAssignableTo(typeof(byte[])))
             {
-                new IonByteArraySerializer().Serialize(writer, (byte[])(object)item);
+                var serializer = this.GetCustomPrimitiveSerializer<byte[]>() ?? new IonByteArraySerializer();
+                serializer.Serialize(writer, (byte[])(object)item);
                 return;
             }
 
-            if (item is string)
+            if (type.IsAssignableTo(typeof(string)))
             {
-                new IonStringSerializer().Serialize(writer, item as string);
+                var serializer = this.GetCustomPrimitiveSerializer<string>() ?? new IonStringSerializer();
+                serializer.Serialize(writer, item as string);
                 return;
             }
 
-            if (item is SymbolToken)
+            if (type.IsAssignableTo(typeof(SymbolToken)))
             {
-                new IonSymbolSerializer().Serialize(writer, (SymbolToken)(object)item);
+                var serializer = this.GetCustomPrimitiveSerializer<SymbolToken>() ?? new IonSymbolSerializer();
+                serializer.Serialize(writer, (SymbolToken)(object)item);
                 return;
             }
 
-            if (item is DateTime)
+            if (type.IsAssignableTo(typeof(DateTime)))
             {
-                new IonDateTimeSerializer().Serialize(writer, (DateTime)(object)item);
+                var serializer = this.GetCustomPrimitiveSerializer<DateTime>() ?? new IonDateTimeSerializer();
+                serializer.Serialize(writer, (DateTime)(object)item);
                 return;
             }
-
-            if (item is System.Collections.IList) 
+            
+            if (type.IsAssignableTo(typeof(Guid)))
             {
-                NewIonListSerializer(item.GetType()).Serialize(writer, (System.Collections.IList)(object)item);
+                var serializer = this.GetCustomPrimitiveSerializer<Guid>() ?? new IonGuidSerializer(options);
+                serializer.Serialize(writer, (Guid)(object)item);
                 return;
             }
 
-            if (item is Guid) 
+            if (type.IsAssignableTo(typeof(System.Collections.IList)))
             {
-                new IonGuidSerializer(options).Serialize(writer, (Guid)(object)item);
+                NewIonListSerializer(type).Serialize(writer, (System.Collections.IList)(object)item);

```

Comment:
So even though the type is assignable to the IList type, we need an interim cast to object?

Reply:
Several of the other types require the interim object cast (won't compile without it) but IList does not need it so I removed it.